docs/tendermint-core/p2p/peer_manager.md

```diff
@@ -0,0 +1,439 @@
+
+#### Errors
+
+The transition fails if:
+
+- the node dialed itself
+- the peer is already in the `Connected` state
+- the node is connected to `MaxConnected` peers, and no slot is suitable for upgrading
+
+Errors are also returned if:
+
+- the dialed peer was removed from the peer store
+- the updated peer information is invalid
+- there is an error when saving the peer state to the peer store
+
+### DialFailed transition
+
+This transition informs a failure when establishing an `outgoing` connection to
+a peer.
+
+The dialed address's `LastDialFailure` time is set, and its `DialFailures`
+counter is increased.
+This information is used to compute the [retry delay](#retry-delay) for the
+dialed address.
+
+The peer manager then spawns a routine that after the computed retry delay
+notifies the next peer to dial routine about the availability of this peer.
+Until then, the peer is the `Frozen Candidate` state.
+
+#### Retry delay
+
+The retry dial is the minimum time, from the latest failed dialing attempt, we
+should wait until dialing a peer address again.
+
+The default delay is defined by `MinRetryTime` parameter.
+If it is set to zero, we *never* retry dialing a peer address.
+
+Upon each failed dial attempt, we increase the delay by `MinRetryTime`, plus an
+optional random jitter of up to `RetryTimeJitter`.
+
+The retry delay should not be longer than the `MaxRetryTime` parameter,
+or `MaxRetryTimePersistent` parameter in the case of persistent peers.
+
+#### Errors
+
+Errors are also returned if:
+
+- the updated peer information is invalid
+- there is an error when saving the peer state to the peer store
+
+### Accepted transition
+
+This transition is performed when the node establishes an *incoming* connection
+with a peer.
+This means that the router has received a connection attempt from this peer and
+successfully established a connection with it.
+
+It may occur, however, that when this transition is invoked the peer is already
+in the [`Connected`](#connected-peer) state.
+The most likely reason is that the router was simultaneously dialing the same
+peer, and has successfully [established a connection](#dialed-transition) with
+it.
+In this case, the transition fails, indicating to the router that is should
+close the accepted connection.
+
+It may also occur that the node is already connected to `MaxConnected` peers,
+which means that all connection slots are full.
+In this case, the peer manager tries to find a connection slot that can be
+[upgraded](#slots-upgrading) to give room for the accepted connection.
+If no suitable connection slot is found, or the hard limit `MaxConnected +
+MaxConnectedUpgrade` of connected peers is reached, the transitions fails.
+
+If the transition succeeds, the peer is set to the
+[`Connected`](#connected-peer) state as an `incoming` peer.
+
+The accepted peer might not be known by the peer manager.
+In this case the peer is registered in the peer store, without any associated
+address.
```

Comment:
Well, if you are connected to it, you _do_ have an address and port, don't you?

Reply:
Yes, we have the address of the other TCP endpoint, which is not necessarily (I would say, not typically) the same address at which the remote peer accept connections...

Reply:
I commented this on the text too.